### PR TITLE
Productionsite performance

### PR DESCRIPTION
**Pull request by hessenfarmer**
just for windows build. Do not merge.

### CODE REVIEW

**Comment by Noordfrees** [comment]
When you open a pull request, you can use the "Create Draft Pull Request" option to ensure nobody will review or merge the PR before you consider it ready

**Comment by hessenfarmer** [comment]
@Nordfrees 
thanks for the hint. I couldn't change it for this branch however

**Comment by hessenfarmer** [comment]
For a reason I don't know this is crashing due to the asser at player.cc:924 siehe travis logs as well.

**Comment by Noordfrees** [comment]
Looking at the logs it might be a similar cause as #3658

**Comment by Noordfrees** [comment]
Or perhaps not. Are you seeing this in your test games as well (steps to reproduce?) or only on travis?

**Comment by hessenfarmer** [comment]
i Have seen this on my linux machine. just take the last v ersion of the branch: 
1. set up a local network game
2. choose map full moon.
3. set all slots to normal ai (personally I assign a different tribe to each slot) 
4. le it run for a couple of minutes

**Comment by Noordfrees** [comment]
After each statistics log() the economy serial is being re-initialized from `ProductionSite::update_actual_statistics()`.
One must never, ever use Game() to access the game, because unlike in Lua this creates a new Game instance independent from the one currently running. Use owner().egbase() instead.

P.S. typo in log: /n -> \n

**Comment by hessenfarmer** [comment]
thanks. didn't know that

**Comment by hessenfarmer** [comment]
Ok this should be ready now to merge.

File: src/logic/map_objects/tribes/productionsite.cc [hunk 1]
```diff
@@ -427,7 +425,8 @@ void ProductionSite::format_statistics_string() {
 	// asked for the statistics string. However this string should only then be constructed.
 
 	// boost::format would treat uint8_t as char
-	const unsigned int percent = get_actual_statistics();
+	const unsigned int percent =
+	   (get_actual_statistics() * 100 / 98 > 100) ? 100 : get_actual_statistics() * 100 / 98;
```
Comment (tibor95) [inline src/logic/map_objects/tribes/productionsite.cc]: std::min(get_actual_statistics() * 100 / 98, 100)
?

**Review by tibor95 — COMMENTED** [review]
It looks good to me, but I left one comment in regard to calculation and clipping to 100

File: src/logic/map_objects/tribes/productionsite.cc (hunk 1 above)
Comment (hessenfarmer) [reply src/logic/map_objects/tribes/productionsite.cc]: should be more efficient. Thanks

File: src/logic/map_objects/tribes/productionsite.cc [hunk 2]
```diff
@@ -1097,10 +1094,10 @@ void ProductionSite::set_default_anim(std::string anim) {
 void ProductionSite::update_actual_statistics(uint32_t duration, const bool produced) {
 	static const uint32_t duration_cap = 180 * 1000;  // This is highest allowed program duration
 	// just for case something went very wrong...
-	static const uint32_t entire_duration = 10 * 60 * 1000;
 	if (duration > duration_cap) {
 		duration = duration_cap;
 	}
+	static const uint32_t entire_duration = 5 * 60 * 1000;
```
Comment (Noordfrees) [inline src/logic/map_objects/tribes/productionsite.cc]: use a `constexpr uint32_t kStatsEntireDuration = ...;` outside the function body

File: data/tribes/buildings/productionsites/atlanteans/smelting_works/init.lua [hunk 3]
```diff
@@ -55,13 +55,26 @@ tribes:new_productionsite_type {
...
-            "call=smelt_iron",
+            "call=smelt_iron_1",
             "call=smelt_gold",
-            "call=smelt_iron",
+            "call=smelt_iron_2",
             "return=no_stats"
          }
       },
-      smelt_iron = {
+      smelt_iron_1 = {
```
Comment (Noordfrees) [inline data/tribes/buildings/productionsites/atlanteans/smelting_works/init.lua:65]: Please add a comment to each of the changed init.lua's explaining why there have to be two identical programs, so this won't be "optimized out" in the future by accident

**Review by Noordfrees — COMMENTED** [review]
Code LGTM, two small comments inline

File: src/logic/map_objects/tribes/productionsite.cc (hunk 2 above)
Comment (hessenfarmer) [reply src/logic/map_objects/tribes/productionsite.cc]: ok but just for my information could you explain this to me? Where shall I incorporate this and why? my c++ knowledge is very limited.

Comment (Noordfrees) [reply src/logic/map_objects/tribes/productionsite.cc]: Just above the function header:
~~~~

constexpr uint32_t kStatsEntireDuration = ...;
constexpr uint32_t kStatsDurationCap = ...;

void ProductionSite::update_actual_statistics(uint32_t duration, const bool produced) {
	if (duration > kStatsDurationCap) {
		duration = kStatsDurationCap;
	}
	const uint32_t past_duration = kStatsEntireDuration - duration;
	...
~~~~

**Review by Noordfrees — APPROVED** [review]

**Pull request merged**